### PR TITLE
Convert prepare package to typescript

### PR DESCRIPTION
**Pull request by motan**
##### Pre-Merge Checklist

- [ ] Tests and/or benchmarks are included
- [ ] Documentation is changed or added
- [ ] Lint process passed (`npm run lint`)
- [ ] Tests passed (`npm run test`)

### CODE REVIEW

**Comment by motan** [comment]
@bigtimebuddy added draft PR to avoid duplication )

**Comment by bigtimebuddy** [comment]
@motan awesome, thanks for the help!

**Comment by cursedcoder** [comment]
It seems there are no changes

File: packages/prepare/src/BasePrepare.ts [hunk 1]
```diff
@@ -1,10 +1,27 @@
-import { Texture, BaseTexture } from '@pixi/core';
+import { Texture, BaseTexture, AbstractRenderer, Renderer } from '@pixi/core';
```
Comment (cursedcoder) [inline packages/prepare/src/BasePrepare.ts]: Not sure if both interfaces should be used here.
Would be nice to leave it to `AbstractRenderer`

Comment (ivanpopelyshev) [reply packages/prepare/src/BasePrepare.ts]: also, maybe `import type` will be better here?

File: packages/prepare/src/BasePrepare.ts [hunk 2]
```diff
@@ -233,7 +259,7 @@ export class BasePrepare
      *          function must return `true` if it was able to add item to the queue.
      * @return {this} Instance of plugin for chaining.
      */
-    registerFindHook(addHook)
+    registerFindHook(addHook: IFindHook): BasePrepare
```
Comment (bigtimebuddy) [inline packages/prepare/src/BasePrepare.ts]:
```suggestion
    registerFindHook(addHook: IFindHook): this
```

File: packages/prepare/src/BasePrepare.ts [hunk 3]
```diff
@@ -250,7 +276,7 @@ export class BasePrepare
      *          function must return `true` if it was able to handle upload of item.
      * @return {this} Instance of plugin for chaining.
      */
-    registerUploadHook(uploadHook)
+    registerUploadHook(uploadHook: IUploadHook): BasePrepare
```
Comment (bigtimebuddy) [inline packages/prepare/src/BasePrepare.ts]:
```suggestion
    registerUploadHook(uploadHook: IUploadHook): this
```

File: packages/prepare/src/BasePrepare.ts [hunk 4]
```diff
@@ -267,7 +293,7 @@ export class BasePrepare
      *        add to the queue
      * @return {this} Instance of plugin for chaining.
      */
-    add(item)
+    add(item: IDisplayObjectExtended | IUploadHook | IFindHook): BasePrepare
```
Comment (bigtimebuddy) [inline packages/prepare/src/BasePrepare.ts]:
```suggestion
    add(item: IDisplayObjectExtended | IUploadHook | IFindHook): this
```

File: packages/prepare/src/BasePrepare.ts [hunk 5]
```diff
@@ -320,7 +346,8 @@ export class BasePrepare
  * @param {Array<*>} queue - Collection of items to upload
  * @return {boolean} if a PIXI.Texture object was found.
  */
-function findMultipleBaseTextures(item, queue)
+
+function findMultipleBaseTextures(item: IDisplayObjectExtended, queue: Array<any>): any
```
Comment (bigtimebuddy) [inline packages/prepare/src/BasePrepare.ts]:
```suggestion
function findMultipleBaseTextures(item: IDisplayObjectExtended, queue: Array<any>): boolean
```

File: packages/prepare/src/BasePrepare.ts [hunk 6]
```diff
@@ -1,10 +1,27 @@
-import { Texture, BaseTexture } from '@pixi/core';
+import { Texture, BaseTexture, AbstractRenderer, Renderer } from '@pixi/core';
 import { Ticker, UPDATE_PRIORITY } from '@pixi/ticker';
 import { settings } from '@pixi/settings';
-import { Container } from '@pixi/display';
+import { Container, DisplayObject } from '@pixi/display';
 import { Text, TextStyle, TextMetrics } from '@pixi/text';
 import { CountLimiter } from './CountLimiter';
+import { CanvasPrepare } from 'pixi.js-legacy';
```
Comment (bigtimebuddy) [inline packages/prepare/src/BasePrepare.ts]: Please don't add CanvasPrepare as a dependency, this is an inappropriate coupling between Prepare and CanvasPrepare. I know it's referenced in the types, but BasePrepare might be fine in its place.

Comment (motan) [reply packages/prepare/src/BasePrepare.ts]: it seems it's added automatically by IDE. Should CanvasPrepare be replaced with BasePrepare or should it be completely removed?

**Comment by bigtimebuddy** [comment]
Thank you @motan!

**Pull request merged**